File: extension/rcas.js
/**
 * RCAS — Robust Contrast-Adaptive Sharpening
 * Based on AMD FidelityFX CAS/RCAS (MIT licensed).
 *
 * Single-pass edge-aware sharpening that enhances texture detail
 * without creating halos around edges. Runs as a post-process
 * after SR upscaling.
 *
 * The algorithm:
 * 1. Sample center pixel + 4 cardinal neighbors
 * 2. Compute local contrast (max - min of the 5 samples)
 * 3. Derive a sharpening weight that's inversely proportional to contrast
 *    (sharpen flat textures more, edges less)
 * 4. Apply a negative-lobe filter clamped to prevent ringing
 */

class RCASPass {
  constructor() {
    this.device = null;
    this.pipeline = null;
    this.bindGroupLayout = null;
    this.sharpness = 0.25;
  }

  async init(device) {
    this.device = device;

    const shaderModule = device.createShaderModule({
      code: RCAS_SHADER,
    });

    this.bindGroupLayout = device.createBindGroupLayout({
      entries: [
        { binding: 0, visibility: GPUShaderStage.COMPUTE, buffer: { type: 'read-only-storage' } },
        { binding: 1, visibility: GPUShaderStage.COMPUTE, buffer: { type: 'storage' } },
        { binding: 2, visibility: GPUShaderStage.COMPUTE, buffer: { type: 'uniform' } },
      ],
    });

    this.pipeline = device.createComputePipeline({
      layout: device.createPipelineLayout({
        bindGroupLayouts: [this.bindGroupLayout],
      }),
      compute: {
        module: shaderModule,
        entryPoint: 'main',
      },
    });
  }

  async apply(inputBuffer, outputBuffer, width, height) {
    const uniformData = new Float32Array([this.sharpness]);
    const sizeData = new Uint32Array([width, height]);
    const uniformBuf = this.device.createBuffer({
      size: 16,
      usage: GPUBufferUsage.UNIFORM | GPUBufferUsage.COPY_DST,
    });

    const combined = new ArrayBuffer(16);
    new Uint32Array(combined, 0, 2).set(sizeData);
    new Float32Array(combined, 8, 1).set(uniformData);
    this.device.queue.writeBuffer(uniformBuf, 0, combined);

    const bindGroup = this.device.createBindGroup({
      layout: this.bindGroupLayout,
      entries: [
        { binding: 0, resource: { buffer: inputBuffer } },
        { binding: 1, resource: { buffer: outputBuffer } },
        { binding: 2, resource: { buffer: uniformBuf } },
      ],
    });

    const encoder = this.device.createCommandEncoder();
    const pass = encoder.beginComputePass();
    pass.setPipeline(this.pipeline);
    pass.setBindGroup(0, bindGroup);
    pass.dispatchWorkgroups(
      Math.ceil(width / 16),
      Math.ceil(height / 16),
      1
    );
    pass.end();
    this.device.queue.submit([encoder.finish()]);
    await this.device.queue.onSubmittedWorkDone();

    uniformBuf.destroy();
  }
}

const RCAS_SHADER = `
struct Params {
  width: u32,
  height: u32,
  sharpness: f32,
  _pad: f32,
}

@group(0) @binding(0) var<storage, read> input: array<f32>;
@group(0) @binding(1) var<storage, read_write> output: array<f32>;
@group(0) @binding(2) var<uniform> params: Params;

fn load(ch: u32, x: i32, y: i32) -> f32 {
  let cx = clamp(x, 0, i32(params.width) - 1);
  let cy = clamp(y, 0, i32(params.height) - 1);
  return input[ch * params.width * params.height + u32(cy) * params.width + u32(cx)];
}

@compute @workgroup_size(16, 16)
fn main(@builtin(global_invocation_id) gid: vec3<u32>) {
  let x = i32(gid.x);
  let y = i32(gid.y);

  if (gid.x >= params.width || gid.y >= params.height) {
    return;
  }

  let W = params.width;
  let H = params.height;

  for (var ch = 0u; ch < 3u; ch++) {
    // 5-tap pattern: center + N/S/E/W
    let c = load(ch, x, y);
    let n = load(ch, x, y - 1);
    let s = load(ch, x, y + 1);
    let w = load(ch, x - 1, y);
    let e = load(ch, x + 1, y);

    // Local min/max for contrast detection
    let mn = min(min(n, s), min(w, e));
    let mx = max(max(n, s), max(w, e));
    let mn2 = min(mn, c);
    let mx2 = max(mx, c);

    // Peak = distance from center to nearest extreme, normalized by range
    let range = mx2 - mn2;
    let peak = select(0.0, (1.0 - min(mn2, 1.0 - mx2) / range), range > 0.001);

    // Sharpening weight: less sharpening near edges (high contrast)
    // More sharpening in flat areas (low contrast)
    let wt = sqrt(min(peak, 1.0)) * (-1.0 / (4.0 * params.sharpness + 4.0));

    // Apply negative-lobe sharpening filter
    let sum = n * wt + s * wt + w * wt + e * wt;
    let result = (c + sum) / (1.0 + 4.0 * wt);

    // Clamp to prevent ringing beyond local min/max
    let final_val = clamp(result, mn2, mx2);

    output[ch * W * H + gid.y * W + gid.x] = final_val;
  }
}
`;

if (typeof globalThis !== 'undefined') {
  globalThis.RCASPass = RCASPass;
}

File: training/rcas.py
"""
RCAS — Robust Contrast-Adaptive Sharpening (PyTorch)

Based on AMD FidelityFX RCAS. Edge-aware sharpening that enhances
texture detail without halos. Meant as a post-process after SR.

Usage:
    from rcas import rcas_sharpen
    sharpened = rcas_sharpen(sr_output, sharpness=0.25)
"""
import torch
import torch.nn.functional as F


def rcas_sharpen(img, sharpness=0.25):
    """
    Apply RCAS sharpening to a [B, 3, H, W] tensor in [0, 1].

    Args:
        img: Input tensor [B, C, H, W] in [0, 1]
        sharpness: Sharpening strength, 0.0 = max sharpening, 1.0 = no sharpening.
                   Default 0.25 is a good balance for SR output.
    Returns:
        Sharpened tensor, same shape.
    """
    # Pad for border pixels
    padded = F.pad(img, (1, 1, 1, 1), mode='reflect')

    # 5-tap: center + cardinal neighbors
    c = img
    n = padded[:, :, :-2, 1:-1]
    s = padded[:, :, 2:, 1:-1]
    w = padded[:, :, 1:-1, :-2]
    e = padded[:, :, 1:-1, 2:]

    # Local min/max
    mn = torch.min(torch.min(n, s), torch.min(w, e))
    mx = torch.max(torch.max(n, s), torch.max(w, e))
    mn2 = torch.min(mn, c)
    mx2 = torch.max(mx, c)

    # Contrast-adaptive weight
    rng = mx2 - mn2
    peak = torch.where(
        rng > 1e-3,
        (1.0 - torch.min(mn2, 1.0 - mx2) / rng),
        torch.zeros_like(rng),
    )
    wt = torch.sqrt(torch.clamp(peak, 0, 1)) * (-1.0 / (4.0 * sharpness + 4.0))

    # Apply sharpening filter
    result = (c + (n + s + w + e) * wt) / (1.0 + 4.0 * wt)

    # Clamp to local range to prevent ringing
    result = torch.clamp(result, mn2, mx2)

    return result


if __name__ == '__main__':
    import os
    import argparse
    from PIL import Image
    from torchvision import transforms

    parser = argparse.ArgumentParser()
    parser.add_argument('input', help='Input image path')
    parser.add_argument('--sharpness', type=float, default=0.25)
    parser.add_argument('--output', default=None)
    args = parser.parse_args()

    img = transforms.ToTensor()(Image.open(args.input).convert('RGB')).unsqueeze(0)
    result = rcas_sharpen(img, args.sharpness)
    out_img = transforms.ToPILImage()(result.squeeze(0))

    out_path = args.output or args.input.replace('.', '_rcas.')
    out_img.save(out_path)
    print(f"Saved: {out_path}")
